inventory.lua

```diff
@@ -1,31 +1,50 @@
-local function sendMessage(pos, msg, channel)
-	if channel == nil then
-		channel = minetest.get_meta(pos):get_string("channel")
-	end
+-- Sends a message onto the Digilines network.
+-- pos: the position of the Digilines chest node.
+-- action: the action string indicating what happened.
+-- stack: the ItemStack that the action acted on (optional).
+-- side: which side of the chest the action occurred (optional).
+local function send_message(pos, action, stack, side)
+	local channel = minetest.get_meta(pos):get_string("channel")
+	local msg = {
+		action = action,
+		stack = stack and stack:to_table(),
+		-- Duplicate the vector in case the caller expects it not to change.
+		side = side and vector.new(side)
+	}
 	digilines.receptor_send(pos,digilines.rules.default,channel,msg)
 end
 
-local function maybeString(stack)
-	if type(stack)=='string' then return stack
-	elseif type(stack)=='table' then return dump(stack)
-	else return stack:to_string()
+-- Checks if the inventory has become empty and, if so, sends an empty message.
+local function check_empty(pos, listname)
+	if minetest.get_meta(pos):get_inventory():is_empty(listname) then
+		send_message(pos, "empty")
 	end
 end
 
-local function can_insert(pos, stack)
-	local can = minetest.get_meta(pos):get_inventory():room_for_item("main", stack)
-	if can then
-		sendMessage(pos,"put "..maybeString(stack))
-	else
-		-- overflow and lost means that items are gonna be out as entities :/
-		sendMessage(pos,"lost "..maybeString(stack))
+-- Checks if the inventory has become full for a particular type of item and,
+-- if so, sends a full message.
+local function check_full(pos, listname, stack)
+	local one_item_stack = ItemStack(stack)
+	one_item_stack:set_count(1)
+	if not minetest.get_meta(pos):get_inventory():room_for_item(listname, one_item_stack) then
+		send_message(pos, "full", one_item_stack)
 	end
-	return can
 end
 
 local tubeconn = minetest.get_modpath("pipeworks") and "^pipeworks_tube_connection_wooden.png" or ""
 local tubescan = minetest.get_modpath("pipeworks") and function(pos) pipeworks.scan_for_tube_objects(pos) end or nil
 
+-- A place to remember things from allow_metadata_inventory_put to
+-- on_metadata_inventory_put.
+local last_inventory_put_listname
+local last_inventory_put_index
+local last_inventory_put_stack
+
+-- A place to remember things from allow_metadata_inventory_take to
+-- tube.remove_items.
+local last_inventory_take_listname
+local last_inventory_take_index
```

Comment:
Er... is the call order guaranteed? E.g. that `on_metadata_inventory_put` will be called after the corresponding `allow_metadata_inventory_put` call but before any other `allow_metadata_inventory_put` call?

Reply:
As far as I can tell yes, though I could be wrong. All the work is done in `src/inventorymanager.cpp` in `IMoveAction::apply`. This function calls `ScriptApiNodemeta::nodemeta_inventory_allowPut` and/or `ScriptApiNodemeta::nodemeta_inventory_allowTake` in the case where the source and destination are not the same inventory, does a pile of calculations and mucking with `ItemStack` objects, and then eventually calls `ScriptApiNodemeta::nodemeta_inventory_OnPut` or `ScriptApiNodemeta::nodemeta_inventory_OnTake`. I tried to find information about threading (i.e. whether there could be a second thread that calls `nodemeta_inventory_allowPut` in between); while there was next to no documentation I could find on how Minetest uses threads, I did find a vague mention somewhere (can’t remember where) that only one thread calls Lua code. Assuming that’s true, and having verified that `IMoveAction::apply` doesn’t AFAICT run any other Lua code in between, then I believe that yes, the call order is guaranteed.

Reply:
That’s not what I meant.
1. Yes, other threads can’t access Lua code/data, it is single-threaded by nature. But
2. The call order is proper *currently*, but won’t it be changed in the future? I wouldn’t be sure. Especially considering the code called by other mods (`pipeworks`).

So, while such hacks might be acceptable currently, they should be marked as such, at least...

Reply:
I wish none of these variables existed. When minetest-mods/pipeworks#205 is fixed, we can drop `allow_metadata_inventory_take` altogether, along with these variables. When the issue opened due to your comment on the other lines further down is closed, the `last_inventory_put` variables can go too. Do you want me to put comments on them pointing at the issues (once both issues exist) and saying they can be deleted once those are fixed?

Reply:
Notes added in 5d8e7009a546.